content/en/blog/_posts/TBD-admission-controllers-for-container-drift/index.md

```diff
@@ -0,0 +1,90 @@
+Once a request of `kubectl exec` comes to our admission controller service, it tags the target pod by adding some customized K8s labels. In this way, we can not only query all the pods with potentially mutated containers, but also enable our service to retrieve previously identified pods in case itself gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object stored in these admission requests is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the labels. The service returns an admission response as allowed after passing that pod reference to its K8s client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected or not, simply by looking at its labels or running a `kubectl describe` command:
+
+```
+$ kubectl get pod --show-labels
+NAME      READY  STATUS   RESTARTS  AGE  LABELS
+test-pod  1/1    Running  0         2s   box.com/podInitialInteractionTimestamp=1632524400,box.com/podInteractorUsername=username-1,box.com/podTTLDuration=1h0m0s
+
+$ kubectl describe pod test-pod
+...
+Events:
+Type       Reason            Age     From                            Message
+----       ------            ----    ----                            -------
+Warning    PodInteraction    5s      admission-controller-service    Pod was interacted with 'kubectl exec' command by user 'username-1' initially at time 2021-09-24 16:00:00 -0800 PST
+Warning    PodInteraction    5s      admission-controller-service    Pod will be evicted at time 2021-09-24 17:00:00 -0800 PST (in about 1h0m0s).
+```
+
+### 3. Evict the target pod after a predefined period
+As you can see in the above event messages, the affected pod is not evicted immediately. At times, developers might have to get into their running containers necessarily for debugging some live issues. Therefore, we define a TTL of affected pods based on the environment of clusters they are running. In particular, we allow a longer time in our dev clusters as it is more common to run `kubectl exec` or other interactive commands for active development. While in prod clusters, there is a more limited time specified to avoid the affected pods serving traffic long term. The controller service internally sets a timer to each pod according to the given TTL duration. Once the timer is up, it will use the K8s Eviction API to restart the pod. The Eviction API ensures service availability since it respects the [PodDisruptionBudget](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) (PDB) corresponding to the Deployment object. Let's say if a user has defined X number of pods as critical in their PDB object, the eviction process, requested by our admission controller service, will not continue when the Deployment has fewer than X pods running.
```

Comment:
```suggestion
 As you can see in the above event messages, the affected pod is not evicted immediately. At times, developers might have to get into their running containers necessarily for debugging some live issues. Therefore, we define a time to live (TTL) of affected pods based on the environment of clusters they are running. In particular, we allow a longer time in our dev clusters as it is more common to run `kubectl exec` or other interactive commands for active development.
 
For our production clusters, we specifc a lower time limit so as to avoid the affected pods serving traffic long term. The kube-exec-controller internally sets and tracks a timer for each pod that matches to the associated TTL. Once the timer is up, the kube-exec-controller evicts that Pod. Using Kubernetes API to request eviction ensures service availability, since the cluster respects any configured [PodDisruptionBudget](/docs/concepts/workloads/pods/disruptions/) (PDB) when performing the eviction. Let's say if a user has defined _x_ number of pods as critical in their PDB, the eviction (as requested by kube-exec-controller) does not continue when the target workload has fewer than _x_ pods running.
```

Reply:
Updated based on the suggestion above.